From 89dd033c4df691eb378a085de175cc6faa31e78e Mon Sep 17 00:00:00 2001
From: Nathan Chancellor <nathan@kernel.org>
Date: Wed, 28 Sep 2022 15:11:37 -0700
Subject: [PATCH] x86/Kconfig: Check for ms_abi attribute support for
 CONFIG_EFI_STUB

Rather than checking for '-mabi=ms', which is not supported by clang,
check that the 'ms_abi' attribute is available.

Cc: stable@vger.kernel.org
Fixes: 8f24f8c2fc82 ("efi/libstub: Annotate firmware routines as __efiapi")
Link: https://github.com/ClangBuiltLinux/linux/issues/1725
Signed-off-by: Nathan Chancellor <nathan@kernel.org>
---
 arch/x86/Kconfig | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index f9920f1341c8..a7b9dce46cd8 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -1953,10 +1953,13 @@ config EFI
 	  resultant kernel should continue to boot on existing non-EFI
 	  platforms.
 
+config CC_HAS_MS_ABI_FN_ATTR
+	def_bool $(success,echo '__attribute__((ms_abi)) int x(void);' | $(CC) $(CLANG_FLAGS) -x c - -c -o /dev/null -Werror)
+
 config EFI_STUB
 	bool "EFI stub support"
 	depends on EFI
-	depends on $(cc-option,-mabi=ms) || X86_32
+	depends on CC_HAS_MS_ABI_FN_ATTR || X86_32
 	select RELOCATABLE
 	help
 	  This kernel feature allows a bzImage to be loaded directly
-- 
2.37.3

